File: backend/app/api/analytics.py
"""Analytics over the incident history: sessions, scoring, geometry, trend.

Every route here is a GET that only reads. The window is either a look-back in
hours or a single ``session_id``; when both are given the session wins, because
"this run" is a more specific question than "the last eight hours".
"""
from __future__ import annotations

import math
import time

from fastapi import APIRouter, HTTPException, Query
from fastapi.concurrency import run_in_threadpool

from ..analytics import reliability
from ..pipeline.session import manager
from ..pipeline.types import CLASS_LABELS
from ..store.db import get_db

router = APIRouter(prefix="/api/analytics", tags=["analytics"])

# Buckets for the lateral-position histogram: the belt's width, in twentieths.
# Twenty is enough to see a defect population sitting off to one side without
# being so fine that a handful of incidents looks like structure.
LATERAL_BINS = 20
# Log-spaced area bins: defect sizes span orders of magnitude, so linear bins
# would put almost everything in the first one.
AREA_BINS = (0.0, 0.0005, 0.002, 0.008, 0.02, 0.04, 0.1, 1.0)


def _live_frames(hours: float | None, session_id: int | None) -> dict[str, int]:
    """In-flight frame counts for the session currently running, if it is in view.

    A session's frame accounting is written to its row only when it *ends*
    (``end_session`` in pipeline/session.py), so the run in progress contributes
    zeroes to any SQL total. Without this the analytics page reports no
    inspection coverage at all for the very run someone is watching, which is
    the run they are most likely to be asking about.

    Adding is safe rather than double-counting: the live session's row still
    holds the column defaults until it closes.
    """
    live = manager.session
    if live is None or not live.running or live.session_id is None:
        return {}
    if session_id is not None and live.session_id != session_id:
        return {}
    status = live.status()
    if session_id is None and hours is not None:
        started = float(status.get("started_at") or 0.0)
        if started < time.time() - hours * 3600:
            return {}

    return {
        "frames_read": int(status.get("frames_read") or 0),
        "frames_processed": int(status.get("frames_processed") or 0),
        "frames_skipped": int(status.get("frames_skipped") or 0),
    }


def _window_bounds(hours: float | None, session: dict | None) -> tuple[float, float]:
    """Start and end of the window, in epoch seconds."""
    now = time.time()
    if session is not None:
        return (float(session["started_at"]),
                float(session.get("ended_at") or now))
    if hours is None:
        return (0.0, now)
    return (now - hours * 3600, now)


@router.get("/sessions")
async def list_sessions(limit: int = Query(50, ge=1, le=500),
                        offset: int = Query(0, ge=0)) -> dict:
    """The run ledger, newest first, with each run's incident tally."""
    db = get_db()
    rows = await run_in_threadpool(db.list_sessions, limit, offset)
    live = manager.session
    live_id = live.session_id if live is not None and live.running else None
    for r in rows:
        # A run with no ended_at is either in progress or was cut off by a
        # crash; the two look identical in the table and mean different things.
        r["is_live"] = (live_id is not None and r["id"] == live_id)
    return {"items": rows, "limit": limit, "offset": offset, "live_id": live_id}


@router.get("/reliability")
async def reliability_index(
    hours: float | None = Query(8.0, ge=0.25, le=8760),
    session_id: int | None = None,
) -> dict:
    """Reliability Yield for one window: condition, coverage, trend.

    The corroboration term is deliberately absent from this payload -- the
    ESP32 node writes to a Supabase project this process holds no credentials
    for, so the two channels are joined in the browser. See
    ``frontend/src/lib/analytics.ts``.
    """
    db = get_db()
    session = None
    if session_id is not None:
        session = await run_in_threadpool(db.get_session, session_id)
        if session is None:
            raise HTTPException(status_code=404, detail="Session not found")

    rows = await run_in_threadpool(db.analytics_rows, hours, session_id)
    frames = await run_in_threadpool(db.frame_totals, hours, session_id)
    for key, value in _live_frames(hours, session_id).items():
        frames[key] = int(frames.get(key) or 0) + value
    start, end = _window_bounds(hours, session)

    # Wall-clock hours are the honest denominator this side of the fusion: the
    # backend cannot see the vibration trace, so it cannot know how much of the
    # window the belt actually spent moving. The browser re-normalises against
    # running time once it has the sensor history.
    observed_hours = max(0.0, end - start) / 3600.0

    payload = reliability.assess(
        rows,
        observed_hours=observed_hours,
        window_start=start,
        window_end=end,
        frames_read=int(frames.get("frames_read") or 0),
        frames_processed=int(frames.get("frames_processed") or 0),
    )
    payload["window"] = {
        "start": start, "end": end, "hours": hours, "session_id": session_id,
    }
    payload["frames"] = {k: int(frames.get(k) or 0) for k in
                         ("frames_read", "frames_processed", "frames_skipped")}
    payload["sessions"] = int(frames.get("sessions") or 0)
    return payload


@router.get("/timeseries")
async def timeseries(
    hours: float | None = Query(8.0, ge=0.25, le=8760),
    session_id: int | None = None,
    buckets: int = Query(48, ge=6, le=240),
) -> dict:
    """Incident counts over time, bucketed by severity and class.

    ``buckets`` is a target resolution rather than a row count: the window is
    divided into that many equal slices, and only non-empty slices come back.
    The client fills the gaps, which keeps an idle night from costing hundreds
    of zero rows on the wire.
    """
    db = get_db()
    session = None
    if session_id is not None:
        session = await run_in_threadpool(db.get_session, session_id)
        if session is None:
            raise HTTPException(status_code=404, detail="Session not found")

    start, end = _window_bounds(hours, session)
    span = max(1.0, end - start)
    bucket_seconds = max(1, int(span / buckets))

    rows = await run_in_threadpool(
        db.incident_buckets, bucket_seconds, hours, session_id)
    return {
        "start": start,
        "end": end,
        "bucket_seconds": bucket_seconds,
        "items": rows,
        "classes": CLASS_LABELS,
    }


@router.get("/geometry")
async def geometry(
    hours: float | None = Query(8.0, ge=0.25, le=8760),
    session_id: int | None = None,
) -> dict:
    """Where across the belt defects appear, and how big they are.

    Lateral position is the reason this endpoint exists. Defects clustering at
    one side of the belt is a *mechanical* finding -- tracking drift, or a
    misaligned roller -- that no amount of looking at severity counts will
    reveal, and every incident row has carried the geometry to show it since
    the first version of the schema.
    """
    db = get_db()
    if session_id is not None:
        if await run_in_threadpool(db.get_session, session_id) is None:
            raise HTTPException(status_code=404, detail="Session not found")

    rows = await run_in_threadpool(db.analytics_rows, hours, session_id)

    lateral = [0] * LATERAL_BINS
    area_hist = [0] * (len(AREA_BINS) - 1)
    scatter: list[dict] = []
    longitudinal = 0
    measured = 0

    for r in rows:
        box = r.get("box")
        if not isinstance(box, (list, tuple)) or len(box) != 4:
            continue
        try:
            x1, y1, x2, y2 = (float(v) for v in box)
        except (TypeError, ValueError):
            continue
        w, h = abs(x2 - x1), abs(y2 - y1)
        if w <= 0 or h <= 0:
            continue
        measured += 1

        centre = min(0.999999, max(0.0, (x1 + x2) / 2.0))
        lateral[int(centre * LATERAL_BINS)] += 1

        area = w * h
        for i in range(len(AREA_BINS) - 1):
            if AREA_BINS[i] <= area < AREA_BINS[i + 1]:
                area_hist[i] += 1
                break

        # Matches LONGITUDINAL_ASPECT in pipeline/events.py -- tall and narrow
        # is the rip-through population, and its share is the risk profile.
        if h / w >= 3.0:
            longitudinal += 1

        scatter.append({
            "id": r.get("id"),
            "cls": r.get("cls"),
            "severity": r.get("severity"),
            "confidence": round(float(r.get("confidence") or 0.0), 4),
            "lateral": round(centre, 4),
            "area": round(area, 6),
            "aspect": round(h / w, 3),
            "opened_at": r.get("opened_at"),
        })

    # Mean and spread of lateral position, so the summary can say "clustered"
    # or "spread evenly" from a number rather than from the shape of a bar.
    centres = [s["lateral"] for s in scatter]
    mean = sum(centres) / len(centres) if centres else None
    if centres and len(centres) > 1:
        var = sum((c - mean) ** 2 for c in centres) / len(centres)
        stdev = math.sqrt(var)
    else:
        stdev = None

    return {
        "measured": measured,
        "total_rows": len(rows),
        "lateral_bins": LATERAL_BINS,
        "lateral": lateral,
        "area_edges": list(AREA_BINS),
        "area": area_hist,
        "longitudinal": longitudinal,
        "lateral_mean": None if mean is None else round(mean, 4),
        "lateral_stdev": None if stdev is None else round(stdev, 4),
        "scatter": scatter,
    }

File: frontend/src/lib/analytics.ts
import type {
  DistinctDefect,
  GeometryResponse,
  ReliabilityIndex,
  Severity,
  TimeseriesResponse,
} from "./types";
import type { ReadingSample, RuptureEvent } from "./rig/useBeltHistory";

/**
 * Where the two sensing channels meet.
 *
 * The camera pipeline and the ESP32 belt-monitor node watch the *same physical
 * belt*, which is what makes everything in this file legitimate: comparing
 * them is comparing two instruments trained on one subject, not conflating two
 * unrelated readings. What must never be conflated is their *verdicts* -- the
 * node's NORMAL/WARNING is its own onboard LDR threshold, not the pipeline's
 * confirmed, severity-scored incident.
 *
 * This join lives in the browser rather than the backend because the two
 * datasets live in two different Supabase projects, and only the browser holds
 * credentials for the rig's.
 */

/**
 * Tolerance when matching a camera incident against a sensor event.
 *
 * `incidents.opened_at` comes from `time.time()` on the API host; a reading's
 * `created_at` is stamped by Supabase's `now()`. Two clocks, never synchronised
 * to each other, plus a confirmation delay of CONFIRM_FRAMES before an incident
 * even opens. Five seconds is generous enough to survive that and still tight
 * enough that a match means something. It is surfaced in the UI, because a
 * correlation quoted without its tolerance is a correlation you cannot check.
 */
export const CORRELATION_TOLERANCE_MS = 5000;

/**
 * Vibration above which the belt is considered to be running.
 *
 * Matches MOTION_THRESHOLD in lib/rig/conveyor-model.ts, which is what the 3D
 * model already uses to decide whether to animate the belt at all. The two must
 * agree: a page reporting the belt as running while the model shows it stopped
 * is reporting on something other than the belt.
 */
export const MOTION_THRESHOLD = 0.12;

/** Classes that describe a rupture, and so could plausibly let light through. */
const RUPTURE_CLASSES = new Set(["joint_damage", "hole", "tear"]);

/* -- Running time --------------------------------------------------------- */

export interface MotionSummary {
  /** Hours the belt actually spent moving, from the vibration trace. */
  runningHours: number;
  /** Hours covered by the readings we have. */
  observedHours: number;
  /** Fraction of observed time the belt was moving, 0..1. */
  dutyCycle: number;
  /** True when there are no readings to judge from. */
  unknown: boolean;
}

/**
 * How long the belt was actually running, from the vibration trace.
 *
 * This is the correction the vision channel cannot make on its own. Defect
 * rate per wall-clock hour is meaningless if the belt stood still for most of
 * the window; per *running* hour is a figure you can compare between shifts.
 */
export function motionSummary(readings: ReadingSample[]): MotionSummary {
  if (readings.length < 2) {
    return { runningHours: 0, observedHours: 0, dutyCycle: 0, unknown: true };
  }

  let runningMs = 0;
  let observedMs = 0;
  for (let i = 1; i < readings.length; i += 1) {
    const dt = readings[i].t - readings[i - 1].t;
    // A gap far larger than the ~1s post interval means the node was offline,
    // not that the belt ran unobserved. Counting it either way would be a
    // guess, so it is counted as neither.
    if (dt <= 0 || dt > 15000) continue;
    observedMs += dt;
    // Trailing edge: the sample at the end of an interval describes it.
    if (readings[i].vibration > MOTION_THRESHOLD) runningMs += dt;
  }

  const observedHours = observedMs / 3_600_000;
  const runningHours = runningMs / 3_600_000;
  return {
    runningHours,
    observedHours,
    dutyCycle: observedMs > 0 ? runningMs / observedMs : 0,
    unknown: observedMs === 0,
  };
}

/* -- Corroboration -------------------------------------------------------- */

export interface Corroboration {
  /** Rupture-class defects the sensor also flagged, within tolerance. */
  matched: number;
  /** Rupture-class defects the camera saw alone. */
  cameraOnly: number;
  /** Sensor rupture events with no camera defect nearby. */
  sensorOnly: number;
  /** matched / (matched + cameraOnly), as a percentage. Null when nothing to score. */
  agreement: number | null;
  /** True when the rig channel produced no data at all in this window. */
  noSensorData: boolean;
}

/**
 * Agreement between the camera and the optical sensor over one window.
 *
 * Only rupture-class defects are scored. A scratch does not let light through
 * the belt, so the LDR array has no opinion about one, and counting the sensor
 * as having "missed" it would slander a working instrument.
 *
 * `noSensorData` exists because "the sensor agreed with nothing" and "there was
 * no sensor" are completely different findings, and rendering the second as 0%
 * would tell an engineer the hardware disagrees when in fact it was unplugged.
 */
export function corroborate(
  defects: DistinctDefect[],
  events: RuptureEvent[],
  readings: ReadingSample[],
  toleranceMs: number = CORRELATION_TOLERANCE_MS,
): Corroboration {
  const relevant = defects.filter((d) => RUPTURE_CLASSES.has(d.cls));

  if (readings.length === 0 && events.length === 0) {
    return {
      matched: 0,
      cameraOnly: relevant.length,
      sensorOnly: 0,
      agreement: null,
      noSensorData: true,
    };
  }

  const usedEvents = new Set<number>();
  let matched = 0;

  for (const d of relevant) {
    // A defect is "seen" from when it opened to when it was last observed;
    // a sensor event anywhere in that span, plus tolerance at each end, is a
    // match.
    const from = d.first_seen * 1000 - toleranceMs;
    const to = d.last_seen * 1000 + toleranceMs;
    const hit = events.findIndex(
      (e, i) => !usedEvents.has(i) && e.t >= from && e.t <= to,
    );
    if (hit >= 0) {
      usedEvents.add(hit);
      matched += 1;
    }
  }

  const cameraOnly = relevant.length - matched;
  const sensorOnly = events.length - usedEvents.size;
  const scored = matched + cameraOnly;

  return {
    matched,
    cameraOnly,
    sensorOnly,
    agreement: scored > 0 ? Math.round((100 * matched) / scored) : null,
    noSensorData: false,
  };
}

/* -- Belt revolution period ----------------------------------------------- */

export interface PeriodEstimate {
  /** Seconds per revolution, or null when no periodicity stands out. */
  seconds: number | null;
  /** Peak autocorrelation, 0..1. Below ~0.3 the estimate is not worth showing. */
  strength: number;
}

/**
 * Estimate the belt's revolution period from vibration periodicity.
 *
 * A belt is a loop, so anything on it -- the splice above all -- passes the
 * sensor once per revolution and leaves a repeating signature in the
 * accelerometer trace. Autocorrelation finds the lag at which the trace best
 * resembles itself.
 *
 * This is an *estimate* and is labelled as one wherever it is shown. It is
 * offered as supporting evidence for the deduplication heuristic, never as a
 * measured constant.
 */
export function estimatePeriod(
  readings: ReadingSample[],
  minSeconds = 2,
  maxSeconds = 120,
): PeriodEstimate {
  const moving = readings.filter((r) => r.vibration > MOTION_THRESHOLD);
  if (moving.length < 30) return { seconds: null, strength: 0 };

  const mean = moving.reduce((a, r) => a + r.vibration, 0) / moving.length;
  const series = moving.map((r) => r.vibration - mean);

  const variance = series.reduce((a, v) => a + v * v, 0);
  if (variance <= 0) return { seconds: null, strength: 0 };

  // Samples arrive at roughly 1 Hz, so a lag in samples is a lag in seconds.
  const maxLag = Math.min(maxSeconds, Math.floor(series.length / 2));
  let bestLag = 0;
  let best = 0;

  for (let lag = minSeconds; lag <= maxLag; lag += 1) {
    let sum = 0;
    for (let i = 0; i + lag < series.length; i += 1) {
      sum += series[i] * series[i + lag];
    }
    const norm = sum / variance;
    if (norm > best) {
      best = norm;
      bestLag = lag;
    }
  }

  if (best < 0.3 || bestLag === 0) return { seconds: null, strength: best };
  return { seconds: bestLag, strength: Math.min(1, best) };
}

/* -- Chart shaping -------------------------------------------------------- */

export interface FusionPoint {
  t: number;
  vibration: number | null;
  vibrationMax: number | null;
  lightPercent: number | null;
  warning: number;
  incidents: number;
  critical: number;
}

/**
 * One row per time bucket carrying both channels, for the fusion timeline.
 *
 * The two sources are bucketed onto a shared grid rather than plotted on their
 * own axes, because the entire point of the chart is that a viewer can read
 * down a vertical line and see what each instrument said at that moment.
 */
export function buildFusionSeries(
  timeseries: TimeseriesResponse | null,
  readings: Array<ReadingSample & { vibrationMax?: number }>,
  bucketMs: number,
): FusionPoint[] {
  const grid = new Map<number, FusionPoint>();

  const at = (t: number): FusionPoint => {
    const key = Math.floor(t / bucketMs) * bucketMs;
    let point = grid.get(key);
    if (!point) {
      point = {
        t: key,
        vibration: null,
        vibrationMax: null,
        lightPercent: null,
        warning: 0,
        incidents: 0,
        critical: 0,
      };
      grid.set(key, point);
    }
    return point;
  };

  for (const r of readings) {
    const p = at(r.t);
    p.vibration = r.vibration;
    p.vibrationMax = r.vibrationMax ?? r.vibration;
    p.lightPercent = r.lightPercent;
    if (r.status === "WARNING") p.warning = 1;
  }

  for (const item of timeseries?.items ?? []) {
    const p = at(item.bucket * 1000);
    p.incidents += item.n;
    if (item.severity === "critical") p.critical += item.n;
  }

  return [...grid.values()].sort((a, b) => a.t - b.t);
}

/* -- Smart summary -------------------------------------------------------- */

export interface SummaryInput {
  index: ReliabilityIndex | null;
  geometry: GeometryResponse | null;
  motion: MotionSummary;
  corroboration: Corroboration;
  period: PeriodEstimate;
}

/**
 * A short plain-English reading of the window.
 *
 * Every sentence is switched on by a threshold over a value that was actually
 * computed; nothing here is a template with a number dropped into it, and
 * there is no language model involved. When the data does not support a
 * claim the sentence is simply not emitted, so a thin window produces a short
 * summary rather than a confident-sounding empty one.
 */
export function buildSummary({
  index,
  geometry,
  motion,
  corroboration,
  period,
}: SummaryInput): string[] {
  const out: string[] = [];
  if (!index) return out;

  const { distinct_defects: distinct, incident_rows: rows } = index;

  if (distinct === 0) {
    out.push(
      index.frames.frames_read === 0
        ? "No footage was processed in this window, so the belt's condition is unknown — this is not a clean bill of health."
        : "No defects were confirmed in this window.",
    );
  } else {
    const critical = index.distinct_by_severity.critical ?? 0;
    const hours = motion.unknown ? index.observed_hours : motion.runningHours;
    const basis = motion.unknown
      ? `${formatSpan(index.observed_hours)} of monitoring`
      : `${formatSpan(hours)} of belt motion`;
    out.push(
      `Over ${basis}, ${distinct} distinct ${distinct === 1 ? "defect was" : "defects were"} confirmed` +
        (critical > 0
          ? `, of which ${critical} ${critical === 1 ? "is" : "are"} critical.`
          : "."),
    );
  }

  // The recurrence correction is worth stating explicitly: it is the single
  // biggest difference between this page and the raw incident count, and an
  // engineer comparing the two needs to know why they disagree.
  if (rows > distinct && distinct > 0) {
    out.push(
      `The camera logged ${rows} sightings of those ${distinct}; the belt is a loop, so a defect returns past the lens once per revolution.` +
        (period.seconds
          ? ` Vibration periodicity puts that revolution at about ${period.seconds}s.`
          : ""),
    );
  }

  // Lateral clustering is a mechanical finding, and the most actionable thing
  // on the page: it points at the rollers, not at the belt surface.
  if (geometry && geometry.measured >= 5 && geometry.lateral_mean !== null &&
      geometry.lateral_stdev !== null && geometry.lateral_stdev < 0.15) {
    const pct = Math.round(geometry.lateral_mean * 100);
    const side = pct < 40 ? "one edge" : pct > 60 ? "the far edge" : "the centre";
    out.push(
      `Defects cluster around ${pct}% of belt width (${side}), rather than spreading across it — a pattern that usually points at belt tracking or roller alignment rather than at the belt surface itself.`,
    );
  }

  if (corroboration.noSensorData) {
    out.push(
      "The belt-monitor node reported nothing in this window, so these findings rest on the camera alone.",
    );
  } else if (corroboration.agreement !== null) {
    out.push(
      `The optical sensor independently flagged ${corroboration.matched} of ${corroboration.matched + corroboration.cameraOnly} rupture-class defects (${corroboration.agreement}% agreement)` +
        (corroboration.sensorOnly > 0
          ? `, and raised ${corroboration.sensorOnly} further ${corroboration.sensorOnly === 1 ? "event" : "events"} the camera did not confirm.`
          : "."),
    );
  }

  // Coverage last, because it qualifies everything above it.
  if (index.coverage !== null && index.coverage < 85) {
    out.push(
      `Inspection coverage was ${index.coverage}% — ${index.frames.frames_skipped.toLocaleString()} frames arrived while the detector was busy and were never analysed, so some belt passed unseen.`,
    );
  }

  if (index.trend !== null && Math.abs(index.trend) >= 5) {
    out.push(
      index.trend < 0
        ? `Condition deteriorated by ${Math.abs(index.trend)} points across the window.`
        : `Condition improved by ${index.trend} points across the window.`,
    );
  }

  return out;
}

/**
 * A duration in hours, written the way someone would say it.
 *
 * A rig session is minutes long, and `0.0h` reads as "no time at all" when the
 * run was a perfectly good forty seconds.
 */
export function formatSpan(hours: number): string {
  const seconds = hours * 3600;
  if (seconds < 90) return `${Math.round(seconds)}s`;
  if (seconds < 5400) return `${Math.round(seconds / 60)} min`;
  return `${hours.toFixed(1)}h`;
}

/** Severity ordering for stacked charts: worst on top. */
export const STACK_ORDER: Severity[] = ["info", "low", "medium", "high", "critical"];
